source/common/grpc/google_async_client_impl.cc

```diff
@@ -0,0 +1,371 @@
+#include "common/grpc/google_async_client_impl.h"
+
+#include "common/common/empty_string.h"
+#include "common/tracing/http_tracer_impl.h"
+
+#include "grpc++/channel.h"
+#include "grpc++/create_channel.h"
+#include "grpc++/security/credentials.h"
+
+namespace Envoy {
+namespace Grpc {
+
+GoogleAsyncClientImpl::GoogleAsyncClientImpl(Event::Dispatcher& dispatcher, Stats::Scope& scope,
+                                             const envoy::api::v2::GrpcService::GoogleGrpc& config)
+    : dispatcher_(dispatcher), stat_prefix_(config.stat_prefix()), scope_(scope) {
+  // TODO(htuch): add support for SSL, OAuth2, GCP, etc. credentials.
+  std::shared_ptr<grpc::ChannelCredentials> creds = grpc::InsecureChannelCredentials();
+  // We rebuild the channel each time we construct the channel. It appears that the gRPC library is
+  // smart enough to do connection pooling and reuse with identical channel args, so this should
+  // have comparable overhead to what we are doing in Grpc::AsyncClientImpl, i.e. no expensive
+  // new connection implied.
+  std::shared_ptr<grpc::Channel> channel = CreateChannel(config.target_uri(), creds);
+  stub_ = std::make_unique<grpc::GenericStub>(channel);
+  // Initialize client stats.
+  stats_.streams_total_ = &scope_.counter("streams_total");
+  for (uint32_t i = 0; i <= Status::GrpcStatus::MaximumValid; ++i) {
+    stats_.streams_closed_[i] = &scope_.counter(fmt::format("streams_closed_{}", i));
+  }
+}
+
+GoogleAsyncClientImpl::~GoogleAsyncClientImpl() {
+  while (!active_streams_.empty()) {
+    active_streams_.front()->resetStream();
+  }
+}
+
+AsyncRequest* GoogleAsyncClientImpl::send(const Protobuf::MethodDescriptor& service_method,
+                                          const Protobuf::Message& request,
+                                          AsyncRequestCallbacks& callbacks,
+                                          Tracing::Span& parent_span,
+                                          const Optional<std::chrono::milliseconds>& timeout) {
+  auto* const async_request =
```

Comment:
super nit: I would reverse this so that allocation is done into unique_ptr and then store a raw pointer for return.

Reply:
I tried this, but it gets a bit messy with how we later add to the AsyncStreamImpl list; there is an additional dynamic_cast required, so it ends up looking better as is.